### ISSUE TITLE
[SPIKE] How Could we Send the Text Message

### ISSUE THREAD

**Issue by loripusey** [department-of-veterans-affairs/va.gov-team#49074]
## Questions
- How can we send texts in a timely manner without hitting rate limiting
- Could we send directly with twilio like the initiate-checkin lambda does?
- Do we need to respect the VEText opt-in? (and if so how do we get that)
    - Or do we assume that if you text "check in" that you, by default, are consenting to receive a text about this appointment
- What infrastructure do we need to support this?
- Could VEText suppress all other outgoing texts during the check-in window for appointments?

## Tasks
- Perform research that answers the above questions

**Comment by acrollet** [comment]
Notes from working session 11/7 with @brianseek @acrollet @wullaski:

> How can we send texts in a timely manner without hitting rate limiting

* one issue is the question of what is a timely manner? Should we attempt to send links at the start of the check-in window? If there is a delivery delay, will that create confusion for the veteran? If a check-in and a pre-check-in link arrive close to one another, will that create confusion?
* on the topic of timely delivery, info [from twilio](https://support.twilio.com/hc/en-us/articles/223133667-Troubleshooting-message-delay-issues-using-API-Explorer):
> SMS messages sent from Twilio typically arrive at their destination within 5 to 10 seconds depending on the country and the carrier. If a message takes more than 30 seconds, we consider the message to have been "delayed".
> There is a large number of queued messages on the "From" number you are using. If the "From" number already has a large number of messages queued, additional messages you try to send are added to the queue rather than being sent immediately. 

Per [Twilio](https://support.twilio.com/hc/en-us/articles/115002943027-Understanding-Twilio-Rate-Limits-and-Message-Queues), the default rate limit for a short code is 100 message segments per second. This limit will need to be shared with existing VEText traffic, unless we obtain a new short code.
* Using a separate short code could introduce trust issues
* we may wish to consider sending one link at the beginning of the day that would work for all appointments for the day
* we should meet with the VEText team to better understand these issues

**Comment by acrollet** [comment]
@benbrasso-agile6 as we're talking through this, we're realizing that the parameters around when a message goes out make a fairly big difference from the technical perspective. So the question is: When, ideally, should a message go out for a given appointment? Should it be right at the start of the check-in window? If so, how large is the window for receiving the message, especially if a veteran has consecutive appointments? 5 minutes? 10 minutes?

**Comment by acrollet** [comment]
> Do we need to respect the VEText opt-in? (and if so how do we get that)

I think we need to assume that the opt-in should be respected. In general, our discussion is pointing at this process needing to run in VEText, since it already does similar work for pre-checkin, and the expertise and access needed are concentrated there.

**Comment by acrollet** [comment]
> What infrastructure do we need to support this?

This is somewhat TBD, but if we assume that the maximum number of appointments that would need a text message sent in any given 1/2 hour window is about 5000, and we have 5 minutes to send all texts, the sending rate required is about 17 messages/second. It usually takes around 2-3 seconds for CHIP to process an initiate-checkin request, implying that about 50 worker processes would need to be available to process these requests, if the query to grab the appointments is fairly lightweight. (Perhaps a bad assumption since all VistA instances need to be queried) This would also cause some fairly spiky additional load on vista-api.

It would likely be a good idea to do a staged rollout with active infrastructure monitoring to ensure we don't negatively impact current functionality.

**Comment by benbrasso-agile6** [comment]
> @benbrasso-agile6 as we're talking through this, we're realizing that the parameters around when a message goes out make a fairly big difference from the technical perspective. So the question is: When, ideally, should a message go out for a given appointment? Should it be right at the start of the check-in window? If so, how large is the window for receiving the message, especially if a veteran has consecutive appointments? 5 minutes? 10 minutes?

Yes, right at the start of the check-in window, to whatever degree possible. E.g., run whatever checks need to happen first, then send out message as soon as possible thereafter. This is to maximize the amount of time a Veteran has to start check in and finish it. I don't see a huge pro of waiting to send the link until, for example, 5 mins before the actual appointment.

**Comment by acrollet** [comment]
Next steps:

- get more clarity on how many appointments are eligible for check-in at any given time, and what the maximum is
- schedule time with someone from VEText (probably Rob Durkin) to explain this ask and get their perspective on the practicalities

**Comment by benbrasso-agile6** [comment]
Offering an "opt in" to receive check in notifications may smooth out concerns around the "too many texts" perception.

**Comment by acrollet** [comment]
I have a meeting scheduled with @robertdurkin tomorrow to go over these questions.

**Comment by acrollet** [comment]
@brianseek @wullaski and I just met with @robertdurkin to go over our VEText specific questions, notes follow:

#### Rate limiting/throttling:
- rate limits are high, 100 segments/second
- most patients use short code, a few hundred whose carriers don't allow short code messaging use a long code
- VEText has never had throttling issues, but possible that it has encountered Twilio's send queue
- VEText batches jobs so that not too many texts go out at once - precheckin does this as part of appointment reminders

#### Messaging opt-in
- VEText is auto opt-in, only vets who have opted out by sending STOP will be excluded
  - VEText always checks the opt-out list before sending
  - as of 11/10/22 725,000 Vets have opted out, roughly 8% of total ~9mil veterans that vetext has sent messages to

#### Regarding implementation for pushing check-in links to Veterans at the beginning of the check-in window for their appointments
- VEText does a similar thing for post-appt messages for community care appts, it gets a list of vets to whom it has sent appt reminders, it sends a day after the appointment.
- VEText might store a list of appts that reminders have been sent to, ensure status is not canceled, send message with initiate-checkin link.
  - throughput is probably not an issue, getting the logic right for the window in which to send the message could be relatively tricky
- VEText would likely run a batch process once every half-hour/hour that pulled appointments eligible for check-in and pushed messages out

#### Suppressing other messages when a Veteran is in the window for checking in for an appointment
- VEText *might* be able to suppress reminders during the check-in window, but it would be difficult/high effort

**Comment by benbrasso-agile6** [comment]
I was curious if we could create an explicit opt-in within the VA.gov CIE specifically for receiving check-in notifications. Of course, if a Veteran were to opt-out of "all" VEText notifications, then we'd need to adhere to that rule.

**Comment by acrollet** [comment]
Takeaway conclusions:

* this is possible to do
* VEText is almost certainly the right place to do it
  * implementation effort/timeline is unknown but not trivial

@loripusey, @benbrasso-agile6 and I just briefly discussed and we think it would be good to schedule a meeting with UX to go over the findings and regroup on next steps.

**Comment by acrollet** [comment]
Placing in validate, @loripusey could you review and let us know if any clarifications or further detail are needed? Thanks!

**Comment by loripusey** [comment]
Thanks @acrollet ! I think the next step is the design sprint, may wait until after Thanksgiving as we have a shortened sprint.

**Comment by loripusey** [comment]
validated

**Comment by acrollet** [comment]
Further items to research:

* can we determine how often VEText currently is sending messages during a check-in window for a given patient?
* what is the current state of the Twilio send queue? How often is it getting filled, what's the average send delay?

**Comment by loripusey** [comment]
Create a new ticket for the [further research](https://github.com/department-of-veterans-affairs/va.gov-team/issues/50217). @acrollet if you feel we can handle this ticket this sprint, feel free to pull it in
